src/Functions/PartitionByHyperplanes.cpp

```diff
@@ -0,0 +1,338 @@
+        {
+            multiplyTileAMX(
+                nested_vectors_data,
+                nested_normals_data,
+                nested_offsets_data,
+                col_res,
+                input_rows_count,
+                dimension,
+                vectors_data_index,
+                normals_data_index,
+                coordinate_index);
+            return;
+        }
+#endif
+        multiplyTileScalar(
+            nested_vectors_data,
+            nested_normals_data,
+            nested_offsets_data,
+            col_res,
+            input_rows_count,
+            dimension,
+            vectors_data_index,
+            normals_data_index,
+            coordinate_index);
+    }
+
+    template <class ColumnType, class ResultColumnType>
+    static void executeInternal(
+        const ColumnType & nested_vectors_data,
+        const ColumnType & nested_normals_data,
+        const IColumn * nested_offsets_data,
+        size_t input_rows_count,
+        size_t dimension,
+        ResultColumnType & col_res)
+    {
+        for (size_t i = 0; i < input_rows_count; i += tile_size)
+        {
+            for (size_t j = 0; j < input_rows_count; j += tile_size)
+            {
+                for (size_t k = 0; k < dimension; k += tile_size)
+                {
+                    multiplyTile(
+                        nested_vectors_data,
+                        nested_normals_data,
+                        nested_offsets_data,
+                        col_res,
+                        input_rows_count,
+                        dimension,
+                        i,
+                        j,
+                        k);
+                }
+            }
+        }
+    }
+
+    static void checkDimension(const ColumnArray::Offsets & offsets, size_t dimension)
+    {
+        size_t prev_offset = 0;
+        for (const auto offset : offsets) {
+            if (offset - prev_offset != dimension) {
+                throw Exception(ErrorCodes::SIZES_OF_ARRAYS_DONT_MATCH, "All vectors must have equal size");
+            }
+            prev_offset = offset;
+        }
+    }
+
+public:
+    static constexpr auto name = "partitionByHyperplanes";
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionPartitionByHyperplanes>(); }
+
+    String getName() const override { return name; }
+
+    bool isVariadic() const override { return true; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    size_t getNumberOfArguments() const override { return 0; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        FunctionArgumentDescriptors mandatory_args{
+            {"vectors", &isArray<IDataType>, nullptr, "Array"},
+            {"normals", &isArray<IDataType>, nullptr, "Array"},
```

Comment:
It seems like a good idea
What types should we use then? `ColumnFloat32` for vec and `ColumnFloat32` for normal? As far as I understand,  each vector will be stored in one cell of a table, so it seems to me that the type should be `ColumnArray` (with length 1 maybe?)

Could you please help me here? It is a bit confusing to think in columnar database terms for the first time

Reply:
Sorry for the long delay.

After some discussion with @alexey-milovidov and my thoughts, we came up with this variant of input types: `vec` should be `ColumnArray` of `Float32` and should contain a couple of vectors. `normals` should be `ColumnConst(ColumnArray)` (in order to satisfy `input_rows_count`). And the return type should be `ColumnFixedString` with `FixedString` for every input vector.

It is necessary to be able to accept multiple normals in the function. Multiple vectors are not necessary, but in my opinion, it is more convenient.

These is one difficult moment I can't manage to though: How to cast ColumnArray of floats to ColumnFloat32? As far as I understand the definition of ColumnArray, it should be possible to represent the same data as a ColumnFloat32 without any copying. @vdimir, could you please help me with this?

Reply:
> vec should be ColumnArray of Float32 and should contain a couple of vectors

Could you clarify why couple?

`vec :: Array(Float32), normals :: Const(Array(Array(Float32)))`, for example:


```
vec
[1, 0, 0] 
[1, 1, 1]
[2, 1, 0]
```

```
normals = [[1,1,0],[1,0,-1]]
offsets = [1.5, 0.5]
```

result column
````
01
10
01
```

is it correct?


As for manipulating with `ColumnArray` you get data column with `getData` method, it returns IColumn and in case `Array(Float32)` it should be `ColumnFloat32`, so you cast it to concrete type and then slice according to offsets (obtained with `getOffsets`)

Reply:
> Could you clarify why couple? 

Because the code will be almost the same (it will still be some kind of matrix multiplication), but processing multiple vectors in one function call will increase performance. If we multiply two matrices and the second one has only one column, in the AMX case we will utilize only 1/16 of the AMX ALU, because we will multiply 16x16 tile and 16x1 tile instead of two 16x16 tiles.

Reply:
Each vector is represented by array, but column contains multiple rows, so per each batch you may process up to `max_block_size` vectors (which is about 65000 by default)

Reply:
> > As for manipulating with ColumnArray you get data column with getData method, it returns IColumn and in case Array(Float32) it should be ColumnFloat32, so you cast it to concrete type and then slice according to offsets (obtained with getOffsets)
> 
> Sorry, there is still some problems with this. The code I wrote works fine as far as I call it with only one array in vector and normals, for example,
> 
> `SELECT partitionByHyperplanes([toFloat32(2.0), toFloat32(3.0)], [toFloat32(1.0), toFloat32(1.0)]))`
> 
> works fine. However,
> 
> `SELECT partitionByHyperplanes([toFloat32(2.0), toFloat32(3.0)], [[toFloat32(1.0), toFloat32(-1.0)],[toFloat32(-1.0), toFloat32(2.0)]])`
> 
> Fails with an exception
> 
> `<Fatal> : Logical error: 'Bad cast from type DB::ColumnArray to DB::ColumnVector<float>'`
> 
> Actually, all the other issues have been fixed, could you please take one more look

But which is correct? Incorrect should return `BAD_ARGUMENTS` or something like that

Reply:
I would expect that both versions will be able to correctly casted to ColumnFloat32

Reply:
An error from the second case occures lines 416 and 421

Reply:
`[toFloat32(1.0), toFloat32(1.0)]` is `ColumnArray` with `ColumnFloat32` inside and `[[toFloat32(1.0), toFloat32(-1.0)],[toFloat32(-1.0), toFloat32(2.0)]]` is is `ColumnArray` with another `ColumnArray` in `data`